### ISSUE TITLE
Need better error handling on CNI add error

### ISSUE THREAD

**Issue by leodotcloud** [rancher/rancher#9418]
rancher/server: 1.6.6-rc1

Ended up in this weird error state when I restarted the ipsec+sidekick service. Moving veth to host namespace failed.

Container of interest: 3e7425939bf217197a76f7771f72277068e40dacdbaaf4663a0720b2c4b2ae2e

```
root@rh1:~# docker ps | grep 3e7425939bf2
3e7425939bf2        rancher/net:holder               "/.r/r /rancher-entry"   2 hours ago         Up 14 minutes                               r-ipsec-ipsec-2-46fcf82f
root@rh1:~#
```

`network-manager` logs:

```
time="2017-07-20T21:11:35Z" level=info msg="Setting up resolv.conf for ContainerId [3e7425939bf217197a76f7771f72277068e40dacdbaaf4663a0720b2c4b2ae2e]"
time="2017-07-20T21:11:35Z" level=info msg="CNI up" cid=896f19c48beb75795e161425258bd36a2db349ac6f0cd8563b496b599903735c networkMode=ipsec
time="2017-07-20T21:11:35Z" level=info msg="CNI up" cid=3e7425939bf217197a76f7771f72277068e40dacdbaaf4663a0720b2c4b2ae2e networkMode=ipsec
time="2017-07-20T21:11:35Z" level=info msg="CNI up" cid=3a544ff97dd64310957f00e30bb7e051e1581fc04f64151da76b7a75efc04b17 networkMode=ipsec
time="2017-07-20T21:11:35Z" level=info msg="Setting up resolv.conf for ContainerId [7f0edd9c30ee8f1b770e5ddb3ce83355b73efae8b481516e2a39097f610f2d06]"
time="2017-07-20T21:11:35Z" level=info msg="CNI up done" cid=3a544ff97dd64310957f00e30bb7e051e1581fc04f64151da76b7a75efc04b17 networkMode=ipsec result=IP4:{IP:{IP:10.42.36.71 Mask:ffff0000} Gateway:10.42.0.1 Routes:[{Dst:{IP:169.254.169.250 Mask:ffffffff} GW:<nil>} {Dst:{IP:0.0.0.0 Mask:00000000} GW:10.42.0.1}]}, DNS:{Nameservers:[] Domain: Search:[] Options:[]}
time="2017-07-20T21:11:35Z" level=info msg="CNI up done" cid=896f19c48beb75795e161425258bd36a2db349ac6f0cd8563b496b599903735c networkMode=ipsec result=IP4:{IP:{IP:10.42.239.49 Mask:ffff0000} Gateway:10.42.0.1 Routes:[{Dst:{IP:169.254.169.250 Mask:ffffffff} GW:<nil>} {Dst:{IP:0.0.0.0 Mask:00000000} GW:10.42.0.1}]}, DNS:{Nameservers:[] Domain: Search:[] Options:[]}
time="2017-07-20T21:11:35Z" level=info msg="CNI up done" cid=3e7425939bf217197a76f7771f72277068e40dacdbaaf4663a0720b2c4b2ae2e networkMode=ipsec result=IP4:{IP:{IP:10.42.240.184 Mask:ffff0000} Gateway:10.42.0.1 Routes:[{Dst:{IP:169.254.169.250 Mask:ffffffff} GW:<nil>} {Dst:{IP:0.0.0.0 Mask:00000000} GW:10.42.0.1}]}, DNS:{Nameservers:[] Domain: Search:[] Options:[]}
time="2017-07-20T21:11:37Z" level=info msg="Applying new port rules"
net.ipv4.conf.docker0.route_localnet = 1
net.ipv4.conf.docker0.route_localnet = 1
time="2017-07-20T22:01:34Z" level=info msg="Setting up resolv.conf for ContainerId [6f6b0975dae2d46bdb8ed3d6318184371d57eb2bb746302d30b652ab348e0453]"
time="2017-07-20T22:01:34Z" level=info msg="CNI up" cid=6f6b0975dae2d46bdb8ed3d6318184371d57eb2bb746302d30b652ab348e0453 networkMode=managed
net.ipv4.conf.docker0.route_localnet = 1
time="2017-07-20T22:01:35Z" level=info msg="CNI up done" cid=6f6b0975dae2d46bdb8ed3d6318184371d57eb2bb746302d30b652ab348e0453 networkMode=managed result=IP4:{IP:{IP:10.42.43.168 Mask:ffff0000} Gateway:10.42.0.1 Routes:[{Dst:{IP:169.254.169.250 Mask:ffffffff} GW:<nil>} {Dst:{IP:0.0.0.0 Mask:00000000} GW:10.42.0.1}]}, DNS:{Nameservers:[] Domain: Search:[] Options:[]}
time="2017-07-20T23:46:51Z" level=info msg="CNI down" cid=896f19c48beb75795e161425258bd36a2db349ac6f0cd8563b496b599903735c networkMode=ipsec
time="2017-07-20T23:46:53Z" level=info msg="Setting up resolv.conf for ContainerId [5171d1e3f2479a1bf8ff2b363676dc677d4f346bcc80792dba9c0fbd268dc62b]"
time="2017-07-20T23:46:53Z" level=info msg="CNI up" cid=5171d1e3f2479a1bf8ff2b363676dc677d4f346bcc80792dba9c0fbd268dc62b networkMode=ipsec
time="2017-07-20T23:46:53Z" level=info msg="CNI up done" cid=5171d1e3f2479a1bf8ff2b363676dc677d4f346bcc80792dba9c0fbd268dc62b networkMode=ipsec result=IP4:{IP:{IP:10.42.175.186 Mask:ffff0000} Gateway:10.42.0.1 Routes:[{Dst:{IP:169.254.169.250 Mask:ffffffff} GW:<nil>} {Dst:{IP:0.0.0.0 Mask:00000000} GW:10.42.0.1}]}, DNS:{Nameservers:[] Domain: Search:[] Options:[]}
net.ipv4.conf.docker0.route_localnet = 1
time="2017-07-20T23:54:20Z" level=info msg="CNI down" cid=5171d1e3f2479a1bf8ff2b363676dc677d4f346bcc80792dba9c0fbd268dc62b networkMode=ipsec
time="2017-07-20T23:54:20Z" level=info msg="CNI down" cid=3a544ff97dd64310957f00e30bb7e051e1581fc04f64151da76b7a75efc04b17 networkMode=ipsec
time="2017-07-20T23:54:25Z" level=info msg="Applying new port rules"
time="2017-07-20T23:54:35Z" level=info msg="CNI down" cid=3e7425939bf217197a76f7771f72277068e40dacdbaaf4663a0720b2c4b2ae2e networkMode=ipsec
time="2017-07-20T23:54:36Z" level=info msg="Applying new port rules"
time="2017-07-20T23:54:36Z" level=info msg="Setting up resolv.conf for ContainerId [3e7425939bf217197a76f7771f72277068e40dacdbaaf4663a0720b2c4b2ae2e]"
time="2017-07-20T23:54:36Z" level=info msg="CNI up" cid=3e7425939bf217197a76f7771f72277068e40dacdbaaf4663a0720b2c4b2ae2e networkMode=ipsec
time="2017-07-20T23:54:36Z" level=error msg="Failed to evaluate network state for 3e7425939bf217197a76f7771f72277068e40dacdbaaf4663a0720b2c4b2ae2e: failed to move veth to host netns: file exists"
time="2017-07-20T23:54:36Z" level=error msg="Error processing event &docker.APIEvents{Action:\"start\", Type:\"container\", Actor:docker.APIActor{ID:\"3e7425939bf217197a76f7771f72277068e40dacdbaaf4663a0720b2c4b2ae2e\", Attributes:map[string]string{\"name\":\"r-ipsec-ipsec-2-46fcf82f\", \"io.rancher.cni.link_mtu_overhead\":\"0\", \"io.rancher.container.ip\":\"10.42.240.184/16\", \"io.rancher.container.name\":\"ipsec-ipsec-2\", \"io.rancher.network.arpsync\":\"true\", \"io.rancher.service.deployment.unit\":\"f8cbab9e-471e-450e-8cb3-b1058300fb56\", \"io.rancher.sidekicks\":\"router\", \"io.rancher.container.mac_address\":\"02:54:1b:d3:76:d2\", \"io.rancher.project_service.name\":\"ipsec/ipsec\", \"io.rancher.scheduler.global\":\"true\", \"io.rancher.service.hash\":\"66e5a4d058439e6fc53f0d8f3c2272dc2ea8343a\", \"io.rancher.service.requested.host.id\":\"2\", \"io.rancher.cni.network\":\"ipsec\", \"io.rancher.container.system\":\"true\", \"io.rancher.network.macsync\":\"true\", \"io.rancher.project.name\":\"ipsec\", \"image\":\"rancher/net:holder\", \"io.rancher.cni.wait\":\"true\", \"io.rancher.container.uuid\":\"46fcf82f-4a26-45f8-95e2-b9bd4c500d7a\", \"io.rancher.service.launch.config\":\"io.rancher.service.primary.launch.config\", \"io.rancher.stack.name\":\"ipsec\", \"io.rancher.stack_service.name\":\"ipsec/ipsec\"}}, Status:\"start\", ID:\"3e7425939bf217197a76f7771f72277068e40dacdbaaf4663a0720b2c4b2ae2e\", From:\"rancher/net:holder\", Time:1500594876, TimeNano:1500594876610239076}. Error: failed to move veth to host netns: file exists"
time="2017-07-20T23:54:38Z" level=info msg="Evaluating state from retry" cid=3e7425939bf217197a76f7771f72277068e40dacdbaaf4663a0720b2c4b2ae2e count=1
time="2017-07-20T23:54:38Z" level=info msg="CNI up" cid=3e7425939bf217197a76f7771f72277068e40dacdbaaf4663a0720b2c4b2ae2e networkMode=ipsec
time="2017-07-20T23:54:38Z" level=info msg="CNI up done" cid=3e7425939bf217197a76f7771f72277068e40dacdbaaf4663a0720b2c4b2ae2e networkMode=ipsec result=IP4:{IP:{IP:10.42.240.184 Mask:ffff0000} Gateway:10.42.0.1 Routes:[{Dst:{IP:169.254.169.250 Mask:ffffffff} GW:<nil>} {Dst:{IP:0.0.0.0 Mask:00000000} GW:10.42.0.1}]}, DNS:{Nameservers:[] Domain: Search:[] Options:[]}
time="2017-07-20T23:54:38Z" level=info msg="Setting up resolv.conf for ContainerId [7163f5170bb65f6501fd839dc443862f46a117ea8fbc083a42100250062f6c3c]"
time="2017-07-20T23:54:38Z" level=info msg="CNI up" cid=7163f5170bb65f6501fd839dc443862f46a117ea8fbc083a42100250062f6c3c networkMode=ipsec
time="2017-07-20T23:54:38Z" level=info msg="CNI up done" cid=7163f5170bb65f6501fd839dc443862f46a117ea8fbc083a42100250062f6c3c networkMode=ipsec result=IP4:{IP:{IP:10.42.25.99 Mask:ffff0000} Gateway:10.42.0.1 Routes:[{Dst:{IP:169.254.169.250 Mask:ffffffff} GW:<nil>} {Dst:{IP:0.0.0.0 Mask:00000000} GW:10.42.0.1}]}, DNS:{Nameservers:[] Domain: Search:[] Options:[]}
```

```
root@3e7425939bf2:/#
root@3e7425939bf2:/# ip addr
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host
       valid_lft forever preferred_lft forever
2: vethr3e7425939b@eth0: <BROADCAST,MULTICAST> mtu 1500 qdisc noop state DOWN group default
    link/ether aa:df:e2:e8:19:ae brd ff:ff:ff:ff:ff:ff
3: eth0@vethr3e7425939b: <NO-CARRIER,BROADCAST,MULTICAST,UP,M-DOWN> mtu 1500 qdisc noqueue state LOWERLAYERDOWN group default
    link/ether 02:54:1b:d3:76:d2 brd ff:ff:ff:ff:ff:ff
    inet 10.42.240.184/16 scope global eth0
       valid_lft forever preferred_lft forever
root@3e7425939bf2:/#
root@3e7425939bf2:/#
root@3e7425939bf2:/#
root@3e7425939bf2:/#
root@3e7425939bf2:/#
root@3e7425939bf2:/#
root@3e7425939bf2:/#
root@3e7425939bf2:/#
root@3e7425939bf2:/#
```

**Possible fix:**
If CNI add fails, we should probably restart the container or delete it so that healthcheck can kick in to fix things?

**Workaround:** 
Kill the problematic container
`docker rm -f 3e7425939bf2`

**More details:**

As part of the fix for https://github.com/rancher/rancher/issues/7955, we changed the naming convention of the veth end on the host side to include the docker id instead of a random one. When a container restarts, two things have to happen in order: CNI Delete, CNI add. If for some reason, the CNI add happens before the veth is cleaned up on the host, it would cause a CNI failure leaving the other end of the veth inside the container. This would leave the container with NO network connectivity at all as it's not connected to the (docker) bridge.

**Comment by alena1108** [comment]
I hit this issue easily in the following scenario:

1) make entry.sh fail for the container. It goes to a stopped state.
2) Once container gets restarted, it goes to a bad state with all interfaces down and the same ip addr output present in the original comment.

Recreating the container mostly helps. Except that sometimes it doesn't :)

**Comment by leodotcloud** [comment]
Also I am wondering why the container start didn't fail:
```
8/4/2017 2:45:15 PMtime="2017-08-04T21:45:15Z" level=info msg="CNI down" cid=00421bbac02be24433afe3741792ee7d5ae6e06f8110837fd6803cf9b749b635 networkMode=ipsec
8/4/2017 2:45:38 PMtime="2017-08-04T21:45:38Z" level=info msg="Setting up resolv.conf for ContainerId [00421bbac02be24433afe3741792ee7d5ae6e06f8110837fd6803cf9b749b635]"
8/4/2017 2:45:39 PMtime="2017-08-04T21:45:39Z" level=info msg="CNI up" cid=00421bbac02be24433afe3741792ee7d5ae6e06f8110837fd6803cf9b749b635 networkMode=ipsec
8/4/2017 2:45:39 PMtime="2017-08-04T21:45:39Z" level=error msg="Failed to evaluate network state for 00421bbac02be24433afe3741792ee7d5ae6e06f8110837fd6803cf9b749b635: failed to move veth to host netns: file exists"
8/4/2017 2:45:39 PMtime="2017-08-04T21:45:39Z" level=error msg="Error processing event &docker.APIEvents{Action:\"start\", Type:\"container\", Actor:docker.APIActor{ID:\"00421bbac02be24433afe3741792ee7d5ae6e06f8110837fd6803cf9b749b635\", Attributes:map[string]string{\"io.rancher.container.ip\":\"10.42.172.135/16\", \"io.rancher.container.name\":\"left\", \"io.rancher.container.pull_image\":\"always\", \"name\":\"r-left-b83d50bc\", \"image\":\"leodotcloud/swiss-army-knife:dev\", \"io.rancher.cni.network\":\"ipsec\", \"io.rancher.cni.wait\":\"true\", \"io.rancher.container.mac_address\":\"02:3b:68:2a:52:6c\", \"io.rancher.container.uuid\":\"b83d50bc-9ef5-4061-a586-8f72d298aefb\"}}, Status:\"start\", ID:\"00421bbac02be24433afe3741792ee7d5ae6e06f8110837fd6803cf9b749b635\", From:\"leodotcloud/swiss-army-knife:dev\", Time:1501883138, TimeNano:1501883138991582745}. Error: failed to move veth to host netns: file exists"
8/4/2017 2:45:41 PMtime="2017-08-04T21:45:41Z" level=info msg="Evaluating state from retry" cid=00421bbac02be24433afe3741792ee7d5ae6e06f8110837fd6803cf9b749b635 count=1
8/4/2017 2:45:41 PMtime="2017-08-04T21:45:41Z" level=info msg="CNI up" cid=00421bbac02be24433afe3741792ee7d5ae6e06f8110837fd6803cf9b749b635 networkMode=ipsec
8/4/2017 2:45:41 PMtime="2017-08-04T21:45:41Z" level=info msg="CNI up done" cid=00421bbac02be24433afe3741792ee7d5ae6e06f8110837fd6803cf9b749b635 networkMode=ipsec result=IP4:{IP:{IP:10.42.172.135 Mask:ffff0000} Gateway:10.42.0.1 Routes:[{Dst:{IP:169.254.169.250 Mask:ffffffff} GW:<nil>} {Dst:{IP:0.0.0.0 Mask:00000000} GW:10.42.0.1}]}, DNS:{Nameservers:[] Domain: Search:[] Options:[]}
```

Need to investigate this.

**Comment by leodotcloud** [comment]
Fixed in `rancher/net:v0.11.7`

**Comment by sangeethah** [comment]
Closing this issue based on testing from https://github.com/rancher/rancher/issues/9578
